### ISSUE TITLE
share webcam/microphone missing?

### ISSUE THREAD

**Issue by senapetschnig** [SafeExamBrowser/seb-win-refactoring#3]
While testing the new Beta i wanted to test 3.0 with a browser web conferencing tool but the share webcam and share audio options are not existing in config tool anymore.
A short test showed that cam/mic sharing is not possible with 3.0 in the moment.
Is this on purpose?
Regards and thank you for this amazing product!

**Comment by dbuechel** [comment]
Yes, SEB 3.0 only provides a subset of the full functionality from SEB 2.x. SEB 3.2 will ultimately provide all functionality, see also our roadmap: https://safeexambrowser.org/about_roadmap_en.html

I suspect that these features of the browser engine are disabled per default. Can you provide a way to test the issue? If it is a matter of a simple configuration flag for the browser engine, then we might be able to already include this functionality in the first release...

**Comment by senapetschnig** [comment]
Yes they are set to "always ask" 
chrome://settings/content/camera
chrome://settings/content/microphone

You can test it via: https://test.webrtc.org/

**Comment by dbuechel** [comment]
Great, thanks. I shall have a look at it tomorrow.

**Comment by dbuechel** [comment]
There indeed is a flag to enable WebRTC functionality ("enable-media-stream"), it however has severe side effects. It namely completely breaks the keyboard monitoring and all custom keyboard shortcuts of SEB (e.g. `CTRL+Q` to quit). It strangely only happens once I visit the test site you specified, but I don't think it is an issue with the site, rather an upstream problem in the browser engine once the WebRTC functionality is initialized.

Do you happen to have another site where I can test the functionality? If it were to work there (i.e. if it actually is an issue with the implementation of the test site), I could include the functionality in SEB 3.0. Otherwise, I am not comfortable including it with the severe side effects described above and would need to postpone the feature to at least version 3.1 (where it hopefully will be fixed by an update of the browser engine).

**Comment by senapetschnig** [comment]
i could offer you a testroom on our webconferencing solution but we would need to schedule a time, and a way to exchange information which is not public?

**Comment by dbuechel** [comment]
I have meanwhile found a couple of other sites and tested WebRTC with them, and I am afraid the issue persists. I think we have no other option than to wait for an update of the browser engine.

I already assigned the issue to the next feature version 3.1.0 (which is planned to be released in Q3 2020), I will certainly revisit the issue once work on it commences.

**Comment by dbuechel** [comment]
I shall nonetheless try to investigate what the exact problem is. I'll probably be able to do so next week or the week after, I will keep you updated.

**Comment by muratcelik27** [comment]
Does any news  for WEBRTC ? 
@dbuechel

**Comment by dbuechel** [comment]
I hope to get to the root of the issue this week. I'll keep you updated.

**Comment by dbuechel** [comment]
Okay, I have bad news. It appears that the issue is in fact within the browser engine itself, I was able to reproduce the problem in a bare-bone browser implementation without any SEB code. In the worst case, the issue is within Chromium itself (there already is another issue related to touch input which I reported in the Chromium developer group but never got any response).

**Comment by dbuechel** [comment]
I shall see where I can report the problem, but it is now clear that this won't be fixed for SEB 3.0.0.

**Comment by dbuechel** [comment]
Turns out this is a known issue with CEF: [https://bitbucket.org/chromiumembedded/cef/issues/2609/chomium-windows-when-using-own-low-level](https://bitbucket.org/chromiumembedded/cef/issues/2609/chomium-windows-when-using-own-low-level).

**Comment by mherbold** [comment]
Hello Damian - I took a look at your link and see that you reached out to CEF about fixing this issue about 3 weeks ago.  Have they responded to you?  In our use case we do not have our test takers using the keyboard (they speak / use sign language instead) and so we would be fine with Romain's proposal of adding a --disable-input-monitoring flag to stop monitoring for keyboard clicks.

**Comment by dbuechel** [comment]
They have unfortunately not. The command-line solution would be fine for us as well.

**Comment by mherbold** [comment]
Wow - If this is not fixed then this would completely eliminate the use of SEB for all exams that use WebRTC.

Do you have access to the Chromium source code and the ability to compile it?  From the other comments in that same bitbucket thread, it seems like it is possible to comment out the offending code.  At this point, it might be worth considering doing just that.  Romain seems to have done so with a PR here: https://bitbucket.org/chromiumembedded/cef/pull-requests/213/completely-disable-the-webrtc-user-input/diff

This would be a perfect temporary fix until CEF gets their act together.

**Comment by dbuechel** [comment]
> Wow - If this is not fixed then this would completely eliminate the use of SEB for all exams that use WebRTC.

We are building a workaround where you can add a barebone Chromium Browser as third-party application and use the WebRTC there. But yes, this is a major problem.

> Do you have access to the Chromium source code and the ability to compile it?

Unfortunately not, we do not have the capacity nor skillset required to build our own version of CEF & CefSharp.

**Comment by dbuechel** [comment]
We decided to provide a workaround solution until the issue is fixed in CEF. We implemented a barebone browser that can be added as third-party application in a SEB configuration. [Chrominimum](https://github.com/SafeExamBrowser/Chrominimum) is WebRTC-enabled and can thus be used for live-proctoring while the integrated browser of SEB could be used to e.g. perform an exam.

A beta version of Chrominimum is available here: https://github.com/SafeExamBrowser/Chrominimum/releases

**Comment by mherbold** [comment]
Hmm ok I am giving this a try.  What I have done:

1) Downloaded and installed Chrominium
2) Opened up the SEB config tool
3) Went to the Applications tab, and added c:\program files\chrominium\chrominium.exe
4) Added one argument that is https://xxx.xxx.com
5) Ticked the "Autostart" checkbox
6) Saved the config file.

When I run SEB it does fire up chrominium but the website it loads is duckduckgo.com... ?

Also it starts as a window.  Any command line or chrominium config options to make it start full screen or maximized at least?

Thanks for doing this.

**Comment by dbuechel** [comment]
I can't reproduce the behavior with duckduckgo.com (this is the default URL in the app.config file). Could you share your configuration?

> Also it starts as a window. Any command line or chrominium config options to make it start full screen or maximized at least?

Yes, I plan to extend the options and will include a parameter to maximize the window, but I have to deal with some pending issues of SEB 3.0 first.

**Comment by dbuechel** [comment]
And I suggest that we move all Chrominimum-related issues to its repository. You can create an issue, if you'd like.

**Comment by e404r** [comment]
HI  I integrated the electron js app into seb-win-refactoring and it works  great

**Comment by fishfree** [comment]
@e404r Would you pls share your detailed method or software?

**Comment by e404r** [comment]
> @e404r Would you pls share your detailed method or software?

1. Create Webview App using electron js
2. Embedding Your App  In Seb-win-refactoring
3.Opened Up The SEB Config Tool
4.Went To The Applications Tab, and added  Your App
5.Ticked the "Autostart" checkbox
6. Saved the config file
7.Run

**Comment by dbuechel** [comment]
It unfortunately looks like we won't be able to fix this for version 3.1.

**Comment by dbuechel** [comment]
Good news: The new [WebView2](https://docs.microsoft.com/en-us/microsoft-edge/webview2/) using Edge/Chromium doesn't have the keyboard monitoring issues CEF/CefSharp has. However, the new control doesn't yet offer the same functionality as CEF/CefSharp, which is why we won't be able to switch to it in the near future. As soon as WebView2 will support all the functionality we need for SEB, we'll however most likely use it as new browser engine and thus fix this issue.

**Comment by dbuechel** [comment]
We just learned that enabling WebRTC doesn't seem to break the input monitoring anymore. However, before we can definitely enable it in the integrated browser engine we need to make sure that this isn't just a coincidence which might stop working with the next browser update, as for example the respective issue in CEF still is set to "won't fix": https://bitbucket.org/chromiumembedded/cef/issues/2609/chomium-windows-when-using-own-low-level.

The upcoming version 3.2 will either way contain a simplified, separately configurable and WebRTC-enabled browser window running with MSFT's WebView2 which should resolve this issue for most use cases.

**Comment by chrisaut** [comment]
Just to clarify, this WebView2 will be some sort of special additional window like the barebones chromium app workaround? 3.2 will NOT completely switch from CEF to WebView2, thus bypassing the now seemingly no longer blocking issue in CEF? 

Does current master contain this already, can you give an example on how we could try it to configure to see if this would fix the issue for us?

Thanks

**Comment by dbuechel** [comment]
> Just to clarify, this WebView2 will be some sort of special additional window like the barebones chromium app workaround?

Yes, exactly. Though it won't appear as a separate application in SEB, it will be shown as a browser window like all other browser windows, just with very basic functionality.

> 3.2 will NOT completely switch from CEF to WebView2, thus bypassing the now seemingly no longer blocking issue in CEF?

No, the main browser engine will still be CefSharp / CEF.

> Does current master contain this already, can you give an example on how we could try it to configure to see if this would fix the issue for us?

I was just about to start working on the WebRTC-enabled window when I learned that apparently this issue no longer exists in CEF. We now have to first discuss how we will proceed, I will give an update here once we made a decision.

**Comment by dbuechel** [comment]
We have now enabled WebRTC for the integrated browser engine of SEB, everything seems to work fine at the moment. The functionality can be tested with the latest beta build for version 3.2: https://sebdev-let.ethz.ch/api/buildjobs/9w68k4mn2ybhgox6/artifacts/SEB_3.2.0.294_SetupBundle.exe

We furthermore decided to not implement the additional browser window running WebView2, as we already are late in our schedule and as that wouldn't bring any benefit. So the WebView2 will only be used for the new remote proctoring functionality coming with version 3.2.

**Comment by mherbold** [comment]
Hello - thanks for the update!

I gave it a try - WebRTC seems to be working.  However, another part of our
exam system, which depends on regular video playback, is not working.  How
would I begin investigating this?  I suspect it might just be due to
cross-domain issues or something like that (the video stream comes from
Amazon).

-- Marvin

**Comment by dbuechel** [comment]
You can enable the dev tools for the integrated browser, in tab "User Interface" under "Browser Window Toolbar". Note however that Chromium does not support playback of some proprietary codecs (like H.264), so you might need to convert your videos to an open format.

**Comment by chrisaut** [comment]
Some quick feedback, 3.2 seems to be fine for our purposes too, WebRTC working. We are still testing but doesn't seem like that WebView2 window is needed.

**Comment by ScottDellinger** [comment]
@dbuechel I have confirmed SEB 3.2 is working with WebRTC (although, I haven't been able to get screen sharing to work, but I suspect this may be intentional). What I have noticed is that the ConfigKey does not seem to be working with SEB 3.2 The required header (X-SafeExamBrowser-ConfigKeyHash) does not appear to be sent with requests when "sendBrowserExamKey" is set to "true". Can you confirm? (I've tested this with my own integration with SEB as well as with the Moodle integration. I'm able to change settings in the seb file and there are no issues with me starting the quiz/activity.

**Comment by dbuechel** [comment]
Great, good to know. The Configuration Key and Browser Exam Key should work, however be aware of the issues mentioned in #66.

**Comment by danschlet** [comment]
@ScottDellinger, hope the issue with the ConfigKey in SEB 3.2 can be resolved for you. We are working on a replacement for sending the ConfigKey in request headers, because it seems browser vendors increasingly discourage changing HTTP request headers. Unfortunately the modern WebView (WKWebView) in macOS and iOS, which is required for WebRTC support, doesn't allow to add custom headers to HTTP requests at all. SEB 3.0 for iOS and macOS will contain a new SEB JavaScript API, allowing to query the ConfigKey and BrowserExamKey. See [this post](https://github.com/SafeExamBrowser/seb-win-refactoring/issues/84#issuecomment-839954031) for details.

**Comment by fishfree** [comment]
How to use the live proctoring feature with the 3.2 version? I cannot figure it out.
